{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Analysis of variance on protein length, gene length and exon number of duplicated, gained and retained genes in 13 extant trematodes since their most recent common ancestor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to perform this analysis one has to first run OMA standalone and Pyham to create a dataframe called trematode_df. See Ancestral_trematode_genome_vs_extant_trematode_genome.ipynb. Because the trematode_df has only protein length from fasta files one has to obtain gene ids for each species and calculate the gene length in bp from the gff files. Exon numbers per gene can also be obtained from gff files. Here we assume a ready dataframe with each gene of each species categorized as duplicated, gained or retained since trematode ancestor, with gene length, protein length and exon number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import statsmodels.api as sm\n",
    "from statsmodels.formula.api import ols\n",
    "import researchpy as rp\n",
    "import statsmodels.stats.multicomp\n",
    "import scipy.stats as stats\n",
    "from sklearn.preprocessing import normalize\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#Define working directory and load the cumulative table\n",
    "working_dir = 'C:/Users/nzaja/Documents/Onedrive/onedrivedocuments'\n",
    "trematode_df = pd.read_csv(working_dir + \"/proteinlength/\" + \"cumulative_table.txt\", sep=\",\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You might first want to create a boxplot for gene length, protein length or exon number per species before you perform the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plotting boxplots of protein lengths of different classes, change protein_len to gene_len or exon_number\n",
    "mylabels = [\n",
    "\"Atriophallophorus winterbourni\",\n",
    "\"Clonorchis sinensis\",\n",
    "\"Echinostoma caproni\",\n",
    "\"Fasciola hepatica\",\n",
    "\"Opisthorchis felineus\",\n",
    "\"Opisthorchis viverrini\",\n",
    "\"Schistosoma bovis\",\n",
    "\"Schistosoma curassoni\",\n",
    "\"Schistosoma haematobium\",\n",
    "\"Schistosoma japonicum\",\n",
    "\"Schistosoma mansoni\",\n",
    "\"Schistosoma mattheei\",\n",
    "\"Schistostoma margrebowiei\",\n",
    "\"Trichobilharzia regenti\"]\n",
    "fig, ax = plt.subplots(figsize=(12,6))\n",
    "colours = [\"palevioletred\", \"gold\", \"darkturquoise\"]\n",
    "ax = sns.boxplot(x=\"species\", y=\"protein_len\", hue=\"class\", palette = colours, \\\n",
    "                 data=trematode_df, showfliers=False)\n",
    "ax.set_xticklabels(mylabels, rotation=90)\n",
    "plt.ylabel(\"Protein Length\")\n",
    "plt.xlabel(\"Species\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the summary statistics."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trematode_genome_filenames = [\n",
    "\"Atriophallophorus_red3.agouti.run1.all.maker.proteins.fa\",\n",
    "\"Clonorchis_sinensis_GCA_003604175.1_ASM360417v1_protein.fa\",\n",
    "\"Echinostoma_caproni_GCA_900618425.1_Egypt_0011_upd_protein.fa\",\n",
    "\"Fasciola_hepatica_GCA_002763495.2_1.0.allpaths.pg_protein.fa\",\n",
    "\"Opisthorchis_felineus_GCA_004794785.1_ICG_Ofel_1.0_protein.fa\",\n",
    "\"Opisthorchis_viverrini_GCA_001990785.1_1.0.pg.lrna_protein.fa\",\n",
    "\"Schistosoma_bovis_GCA_003958945.1_ASM395894v1_protein.fa\",\n",
    "\"Schistosoma_curassoni_GCA_900618015.1_Dakar_0011_upd_protein.fa\",\n",
    "\"Schistosoma_haematobium_GCA_000699445.1_1.0_protein.fa\",\n",
    "\"Schistosoma_japonicum_GCA_006368765.1_ASM636876v1_protein.fa\",\n",
    "\"Schistosoma_mansoni_GCA_000237925.2_ASM23792v2_protein.fa\",\n",
    "\"Schistosoma_mattheei_GCA_900617995.1_Denwood_0011_upd_protein.fa\",\n",
    "\"Schistostoma_margrebowiei_GCA_900618395.1_Zambia_0011_upd_protein.fa\",\n",
    "\"Trichobilharzia_regenti_GCA_900618515.1_v1_0_4_001_upd_protein.fa\"]\n",
    "\n",
    "\n",
    "trematode_genomes = [x.split(\".\")[0] for x in trematode_genome_filenames] \n",
    "\n",
    "#Change protein_len into gene_len or exon_number\n",
    "for species in trematode_genomes:\n",
    "    data1=trematode_df[(trematode_df['species']== species) & (trematode_df['class']== 'duplicated')]\n",
    "    data1['protein_len']\n",
    "    print(species, len(data1[['protein_len']]))\n",
    "\n",
    "#mean\n",
    "    sum1 = sum(data1['protein_len'])/len(data1['protein_len'])\n",
    "    print(\"Mean \", sum1)\n",
    "# atrio retained 548.7659275891103\n",
    "\n",
    "#median, 412 for retained atrio\n",
    "    n=len(data1['protein_len'])\n",
    "    data2= data1['protein_len'].tolist()\n",
    "    data2.sort() \n",
    "    if n % 2 == 0: \n",
    "        median1 = data2[n//2] \n",
    "        median2 = data2[n//2 - 1] \n",
    "        median = (median1 + median2)/2\n",
    "    else: \n",
    "        median = data2[n//2] \n",
    "    print(\"Median \" + str(median)) \n",
    "\n",
    "##mode, Mode is / are: 239, 369 atrio retained\n",
    "\n",
    "    data_counter = Counter(data2) \n",
    "    get_mode = dict(data_counter) \n",
    "    mode = [k for k, v in get_mode.items() if v == max(list(data_counter.values()))] \n",
    "  \n",
    "    if len(mode) == n: \n",
    "        get_mode = \"No mode found\"\n",
    "    else: \n",
    "        get_mode = \"Mode \" + ', '.join(map(str, mode)) \n",
    "      \n",
    "    print(get_mode) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Make a distribution of protein length, gene length or exon number per species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Change protein_len to exon_number or gene_len\n",
    "def make_size_plot(df, protein_len_cutoff=2000, alpha=0.5, figsize=(12,7), title=\"title goes here\"):\n",
    "    \n",
    "    alpha = alpha\n",
    "    fig, ax = plt.subplots(figsize=figsize)\n",
    "\n",
    "    subset_df = df[df['protein_len']<=protein_len_cutoff]\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Atriophallophorus_red3\"]['protein_len'], color=\"red\", hist=False, label=\"Atriophallophorus winterbourni\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Opisthorchis_felineus_GCA_004794785\"]['protein_len'], color=\"maroon\", hist=False,  label=\"Opisthorchis felineus\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Clonorchis_sinensis_GCA_003604175\"]['protein_len'], color=\"sandybrown\", hist=False,  label=\"Clonorchis sinensis\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Echinostoma_caproni_GCA_900618425\"]['protein_len'], color=\"purple\", hist=False,  label=\"Echinostoma caproni\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Fasciola_hepatica_GCA_002763495\"]['protein_len'], color=\"magenta\", hist=False,  label=\"Fasciola hepatica\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Opisthorchis_viverrini_GCA_001990785\"]['protein_len'], color=\"darksalmon\", hist=False,  label=\"Opisthorchis viverrini\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Schistosoma_bovis_GCA_003958945\"]['protein_len'], color=\"olive\", hist=False,  label=\"Schistosoma bovis\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Schistosoma_curassoni_GCA_900618015\"]['protein_len'], color=\"yellowgreen\", hist=False,   label=\"Schistosoma curassoni\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Schistosoma_haematobium_GCA_000699445\"]['protein_len'], color=\"lawngreen\", hist=False,  label=\"Schistosoma haematobium\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Schistosoma_japonicum_GCA_006368765\"]['protein_len'], color=\"lightgreen\", hist=False,   label=\"Schistosoma japonicum\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Schistosoma_mansoni_GCA_000237925\"]['protein_len'], color=\"g\", hist=False,  label=\"Schistosoma mansoni\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Schistosoma_mattheei_GCA_900617995\"]['protein_len'], color=\"mediumseagreen\", hist=False, label=\"Schistosoma mattheei\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Schistostoma_margrebowiei_GCA_900618395\"]['protein_len'], color=\"mediumturquoise\", hist=False, label=\"Schistosoma margrebowiei\")\n",
    "    ax = sns.distplot(subset_df[subset_df['species']==\"Trichobilharzia_regenti_GCA_900618515\"]['protein_len'], color=\"darkblue\", hist=False,  label=\"Trichobilharzia regenti\")\n",
    "\n",
    "    ax.set_xlabel(\"Protein Length\")\n",
    "    ax.legend(fontsize=12)\n",
    "    ax.set_title(title, fontsize=20)\n",
    "    \n",
    "#all species combined\n",
    "make_size_plot(trematode_df, title=\"All trematode species\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prepare a table for ANOVA analysis. We assume here protein length (protein_len) but it can be changed for gene length or exon number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "protein_length_df = trematode_df[['species','class', 'protein_len']] #change this to gene_len or exon_nb\n",
    "protein_length_2_df = pd.melt(protein_length_df, id_vars=['species', 'class', 'hog'])\n",
    "protein_length_2_df.columns = ['species', 'gene', 'hog','variable', 'value']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Log transform and normalize values if the data has a lot of outliers and is not normally distributed\n",
    "import numpy as np\n",
    "protein_length_2_df = protein_length_2_df.replace(0, np.nan)\n",
    "protein_length_2_df = protein_length_2_df[protein_length_2_df['value'].notna()]\n",
    "protein_length_log = np.log(protein_length_2_df.value)\n",
    "protein_length_log.describe()\n",
    "protein_length_2_df.insert(4, \"LOG_Values\", protein_length_log, True)\n",
    "#normalization\n",
    "def normalize(column):\n",
    "    upper = column.max()\n",
    "    lower = column.min()\n",
    "    y = (column - lower)/(upper-lower)\n",
    "    return y\n",
    "\n",
    "protein_length_log_normalized = normalize(protein_length_log)\n",
    "protein_length_log_normalized.describe()\n",
    "protein_length_2_df.insert(5, \"LOG_NORM_Values\", protein_length_log_normalized, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Obtain summary statistics for the normalized and log transformed values\n",
    "rp.summary_cont(protein_length_2_df['LOG_NORM_Values'])\n",
    "rp.summary_cont(protein_length_2_df.groupby(['gene']))['LOG_NORM_Values']\n",
    "rp.summary_cont(protein_length_2_df.groupby(['species']))['LOG_NORM_Values']\n",
    "rp.summary_cont(protein_length_2_df.groupby(['gene', 'species']))['LOG_NORM_Values']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run the model for all data\n",
    "model = ols('LOG_NORM_Values ~ C(species) + C(gene) + C(species):C(gene)', data= protein_length_2_df).fit()\n",
    "print(f\"Overall model F({model.df_model: .0f},{model.df_resid: .0f}) = {model.fvalue: .3f}, p = {model.f_pvalue: .4f}\")\n",
    "\n",
    "#Obtain the full model summary\n",
    "model.summary()\n",
    "\n",
    "#Obtain the effect sizes for each independent factor\n",
    "res = sm.stats.anova_lm(model, typ= 2)\n",
    "res\n",
    "def anova_table(aov):\n",
    "    aov['mean_sq'] = aov[:]['sum_sq']/aov[:]['df']\n",
    "    \n",
    "    aov['eta_sq'] = aov[:-1]['sum_sq']/sum(aov['sum_sq'])\n",
    "    \n",
    "    aov['omega_sq'] = (aov[:-1]['sum_sq']-(aov[:-1]['df']*aov['mean_sq'][-1]))/(sum(aov['sum_sq'])+aov['mean_sq'][-1])\n",
    "    \n",
    "    cols = ['sum_sq', 'mean_sq', 'df', 'F', 'PR(>F)', 'eta_sq', 'omega_sq']\n",
    "    aov = aov[cols]\n",
    "    return aov\n",
    "anova_table(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Tukey's Posthoc test to test for significant differences between means of pairs of groups of independent factors in the model\n",
    "#First we test gene category (retained, duplicated, gained)\n",
    "mc = statsmodels.stats.multicomp.MultiComparison(protein_length_2_df['value'], protein_length_2_df['gene'])\n",
    "mc_results = mc.tukeyhsd()\n",
    "print(mc_results)\n",
    "\n",
    "#Then we test species (retained, duplicated, gained)\n",
    "mc = statsmodels.stats.multicomp.MultiComparison(protein_length_2_df['value'], protein_length_2_df['species'])\n",
    "mc_results = mc.tukeyhsd()\n",
    "print(mc_results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Run the model for a subset of 24 HOGs with retained genes and 13 with duplicated genes\n",
    "table = trematode_df[trematode_df['class'] != \"gained\"].groupby(['species', 'hog', 'class']).size().groupby(['hog', 'class']).size().reset_index(name = 'species_nb')\n",
    "list_of_hogs_with_all_13species = table[table['species_nb'] == 13]['hog'].tolist()\n",
    "data_24ret_13dup = trematode_df[trematode_df['hog'].isin(list_of_hogs_with_all_13species)]\n",
    "\n",
    "protein_length_df = data_24ret_13dup[['species','class', 'protein_len']] ##change to gene_len or exon_nb\n",
    "protein_length_2_df = pd.melt(protein_length_df, id_vars=['species', 'class', 'hog'])\n",
    "protein_length_2_df.columns = ['species', 'gene', 'hog' 'variable', 'value']\n",
    "\n",
    "import statsmodels.formula.api as smf\n",
    "model2 = smf.mixedlm(\"LOG_NORM_Values ~ C(gene)\", protein_length_2_df, groups= \"species\", vc_formula = {\"hog\" : \"0 + C(hog)\"}).fit()\n",
    "\n",
    "#Obtain the full model summary\n",
    "model2.summary()\n",
    "\n",
    "#Run the model for a subset of x HOGs per species with retained genes and duplicated genes\n",
    "##prepare the table first, this is an example for protein length but all has to be repeated for gene length and exon number\n",
    "table = trematode_df[trematode_df['class'] != \"gained\"].groupby(['species', 'hog', 'class']).size().groupby(['hog', 'species']).size().reset_index(name = 'class_nb')\n",
    "list_of_hogs_with_dupandret_genes_per_species = l[l['class_nb'] == 2]['hog'].tolist()\n",
    "data_retanddup_onehog_perspecies = trematode_df[trematode_df['hog'].isin(list_of_hogs_with_dupandret_genes_per_species)]\n",
    "data_retanddup_onehog_perspecies = data_retanddup_onehog_perspecies.rename({'class': 'gene_type'}, axis = 1)\n",
    "species = ['Atriophallophorus_red3',\n",
    " 'Clonorchis_sinensis_GCA_003604175',\n",
    " 'Echinostoma_caproni_GCA_900618425',\n",
    " 'Fasciola_hepatica_GCA_002763495',\n",
    " 'Opisthorchis_felineus_GCA_004794785',\n",
    " 'Opisthorchis_viverrini_GCA_001990785',\n",
    " 'Schistosoma_bovis_GCA_003958945',\n",
    " 'Schistosoma_curassoni_GCA_900618015',\n",
    " 'Schistosoma_haematobium_GCA_000699445',\n",
    " 'Schistosoma_japonicum_GCA_006368765',\n",
    " 'Schistostoma_margrebowiei_GCA_900618395',\n",
    " 'Schistosoma_mattheei_GCA_900617995',\n",
    " 'Trichobilharzia_regenti_GCA_900618515']\n",
    "hogs = data_retanddup_onehog_perspecies.hogs.unique().tolist()\n",
    "gene_types = ['retained', 'duplicated']\n",
    "\n",
    "for s in species:\n",
    "    for hog in hogs:\n",
    "        for gene_type in gene_types:\n",
    "            data = table1[(table1['species'] == s) & (table1['hog'] == hog) & (table1['gene_type'] == gene_type)]\n",
    "            if len(data) != 0:\n",
    "                l = statistics.mean(data['protein_len'])\n",
    "                df = pd.DataFrame([[s, hog, gene_type, l]], columns = list('ABCD'))\n",
    "                output = pd.concat([output, df])\n",
    "                \n",
    "output = output.rename({'A' : 'species', 'B' : 'hog', 'C': 'gene_type', 'D' : 'protein_len_average'}, axis = 1)\n",
    "output2 = pd.pivot_table(output, index=['species', 'hog'],columns='gene_type', values = 'protein_len_average', aggfunc='first').reset_index()\n",
    "output3 = output2[(output2['duplicated'].notna()) & (output2['retained'].notna())]\n",
    "output3['retained-duplicated'] = output3['retained'] - output3['duplicated']\n",
    "##count the number of HOGs in each category\n",
    "output4 = pd.DataFrame(columns = [\"A\", \"B\"])\n",
    "for s in species:\n",
    "    data = output3[(output3['species'] == s) & (output3['retained-duplicated'] > 0)]['retained-duplicated'].count()\n",
    "    df = pd.DataFrame([[s, data]], columns = list('AB'))\n",
    "    output4 = pd.concat([output4, df])\n",
    "output4 = output4.rename({'A' : 'species', 'B' : 'retained-duplicated>0'}, axis = 1)\n",
    "\n",
    "output5 = pd.DataFrame(columns = [\"A\", \"B\"])\n",
    "for s in species:\n",
    "    data = output3[(output3['species'] == s) & (output3['retained-duplicated'] < 0)]['retained-duplicated'].count()\n",
    "    df = pd.DataFrame([[s, data]], columns = list('AB'))\n",
    "    output5 = pd.concat([output5, df])\n",
    "output5 = output5.rename({'A' : 'species', 'B' : 'retained-duplicated<0'}, axis = 1)\n",
    "\n",
    "output6 = pd.DataFrame(columns = [\"A\", \"B\"])\n",
    "for s in species:\n",
    "    data = output3[(output3['species'] == s) & (output3['retained-duplicated'] == 0)]['retained-duplicated'].count()\n",
    "    df = pd.DataFrame([[s, data]], columns = list('AB'))\n",
    "    output6 = pd.concat([output5, df])\n",
    "output6 = output6.rename({'A' : 'species', 'B' : 'retained-duplicated=0'}, axis = 1)\n",
    "\n",
    "dfs = [output4, output5, output6]\n",
    "dfs = [df.set_index('species') for df in dfs]\n",
    "protein_length_table = dfs[0].join(dfs[1:]).reset_index()\n",
    "#remove all outputX tables\n",
    "\n",
    "###calculate the proportions\n",
    "protein_length_table['sum'] = protein_length_table['retained-duplicated>0'] + protein_length_table['retained-duplicated<0'] + protein_length_table['retained-duplicated=0']\n",
    "protein_length_table['%retained-duplicated>0'] = (protein_length_table['retained-duplicated>0']/protein_length_table['sum'])*100\n",
    "protein_length_table['%retained-duplicated<0'] = (protein_length_table['retained-duplicated<0']/protein_length_table['sum'])*100\n",
    "protein_length_table['%retained-duplicated=0'] = (protein_length_table['retained-duplicated=0']/protein_length_table['sum'])*100\n",
    "protein_length_table = protein_length_table[['species','%retained-duplicated>0','%retained-duplicated<0', '%retained-duplicated=0']]\n",
    "protein_length_2_df = pd.melt(protein_length_table, id_vars = \"species\")\n",
    "##fit the model\n",
    "model3 = smf.mixedlm(\"value ~ C(variable)\", protein_length_2_df, groups= \"species\", vc_formula = {\"variable\" : \"0 + C(variable)\"}).fit()\n",
    "model3.summary()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
